### PR TITLE
Aarondonahue/scorpio integration interface phase2 input

### PR DESCRIPTION
**Pull request by AaronDonahue**
Extension of the scream-scorpio interface to include input.
    
    This commit extends the framework already in place for accessing
    scorpio output routines to now include access to scorpio input routines.
    
    The set of unit tests is increased by one, thus a file is created and
    has output written to it in test #1.
    
    Test #2 opens that file and reads all fields in as input, testing against
    the expected field values.
    
    The testing has also been extended to include up to 4 threads.
    
    [BFB]

### CODE REVIEW

**Comment by AaronDonahue** [comment]
@brhillman , things moved faster than I expected.  This PR should provide the ability to get input from a netcdf file in scream.

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 [hunk 1]
```diff
@@ -415,6 +445,8 @@ subroutine eam_init_pio_subsystem(mpicom,atm_id,local)
     integer :: num_aggregator
     integer :: base
 
+    if (associated(pio_subsystem)) call errorHandle("PIO ERROR: pio_subsystem has already been defined.",-999)
```
Comment (bartgol) [inline components/scream/src/mct_coupling/scream_scorpio_interface.F90:448]: Maybe I don't understand what 'sync-ing' and pio_subsystem mean, but...isn't the PIO subsystem expected to be already defined at when you try to sync a file?

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 [hunk 2]
```diff
@@ -450,13 +481,35 @@ subroutine eam_pio_createfile(File,fname)
...
+  subroutine eam_pio_openfile(File,fname)
+
+    type(file_desc_t), intent(inout) :: File             ! Pio file Handle
+    character(len=*),  intent(in)    :: fname            ! Pio file name
+    !--
+    integer                          :: retval           ! PIO error return value
+    integer                          :: mode             ! Mode for how to handle the new file
+
+    mode = 0 ! TODO: make sure this is correct. 
+    retval = pio_openfile(pio_subsystem,File,pio_iotype,fname,mode) ! TODO, add error check to use retval, also for createfile below
```
Comment (bartgol) [inline components/scream/src/mct_coupling/scream_scorpio_interface.F90:494]: I think it would be good to add such check to this pr. It doesn't have to go through all possible error codes, just check whether retval signals an error (of any kind) or not.

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 [hunk 3]
```diff
@@ -601,23 +654,27 @@ subroutine get_var(pio_file,varname,var)
...
-  subroutine get_pio_atm_file(filename,pio_atm_file,found)
+  subroutine get_pio_atm_file(filename,pio_file,found,purpose)
 
-    character(len=*),intent(in)   :: filename     ! Name of file to be found
-    type(pio_atm_output), pointer :: pio_atm_file ! Pointer to pio_atm_output structure associated with this filename
-    logical,optional,intent(out)  :: found        ! return TRUE if pio file already exists
+    character(len=*),intent(in)  :: filename     ! Name of file to be found
+    type(pio_atm_file), pointer  :: pio_file     ! Pointer to pio_atm_output structure associated with this filename
+    logical,optional,intent(out) :: found        ! return TRUE if pio file already exists
+    integer,optional,intent(in)  :: purpose      ! Purpose for this file lookup, 0 = as output, 1 = as input
```
Comment (bartgol) [inline components/scream/src/mct_coupling/scream_scorpio_interface.F90:662]: Since these interfaces are likely going to be used only inside scream infrastructure/AD stuff, let's choose safety over ease: let's make purpose a _mandatory_ argument, so we avoid "assuming" any intent.

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 (hunk 1 above)
Comment (AaronDonahue) [reply components/scream/src/mct_coupling/scream_scorpio_interface.F90:448]: Ahh yes, this is the local pointer to the subsystem that has already been defined.  So it is really a matter of whether or not the module level pointer has been assigned.  Perhaps a better error message would be something like "PIO ERROR: The pio_subsystem pointer for this component has already been established", I'll give it some thought on a better message, that too may be a bit too generic.

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 [hunk 4]
```diff
@@ -630,9 +687,16 @@ subroutine get_pio_atm_file(filename,pio_atm_file,found)
...
     curr%pio_file%filename = trim(filename)
-    curr%pio_file%numRecs = 0
-    call eam_pio_createfile(curr%pio_file%pioFileDesc,trim(curr%pio_file%filename))
-    pio_atm_file => curr%pio_file  
+    if (purpose_loc == 0) then
+      curr%pio_file%numRecs = 0
+      call eam_pio_createfile(curr%pio_file%pioFileDesc,trim(curr%pio_file%filename))
+    elseif (purpose_loc == 1) then
+      call eam_pio_openfile(curr%pio_file%pioFileDesc,trim(curr%pio_file%filename))
+      found = .true.
```
Comment (bartgol) [inline components/scream/src/mct_coupling/scream_scorpio_interface.F90:695]: This is where my F90 skills betray me: shouldn't this assignment be guarded by a `if(present(found))` clause?

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 (hunk 3 above)
Comment (AaronDonahue) [reply components/scream/src/mct_coupling/scream_scorpio_interface.F90:662]: That's a good point.  I can make that change.

File: components/scream/src/mct_coupling/tests/CMakeLists.txt [hunk 5]
```diff
@@ -19,15 +19,12 @@ set (SCREAM_INCLUDE
...
-message("ASD")
-message("${ARRAY_SCORPIO_SRCS}")
-message("${SCREAM_LIBS}")
-message("${SCREAM_INCLUDE}")
 # Test atmosphere processes
 EkatCreateUnitTest(scorpio_interface "${ARRAY_SCORPIO_SRCS}" 
   LIBS ${SCREAM_LIBS} csm_share
   LIBS_DIRS ${SCREAM_INCLUDE} ${CSM_SHARE}
   INCLUDE_DIRS ${SCREAM_INCLUDE}
+  THREADS 1 4
```
Comment (bartgol) [inline components/scream/src/mct_coupling/tests/CMakeLists.txt:27]: I don't think we want to run I/O threaded. Perhaps you meant to use MPI_RANK?

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 (hunk 4 above)
Comment (AaronDonahue) [reply components/scream/src/mct_coupling/scream_scorpio_interface.F90:695]: I think you are correct.  I guess I didn't have any cases where I was not passing the optional argument of `found`

File: components/scream/src/mct_coupling/tests/CMakeLists.txt (hunk 5 above)
Comment (AaronDonahue) [reply components/scream/src/mct_coupling/tests/CMakeLists.txt:27]: yes, you are right.  I'll make that change.  Is there anything special that I have to do in CMAKE to make sure that MPI is enabled for testing?  I see in the build log that openmp is enabled but I'm not sure that mpicxx is being used.

**Comment by AaronDonahue** [comment]
It looks like the Autotester got to merging this PR faster than I could make some adjustments based on @bartgol 's comments.  I can weave those changes into my next PR.

File: components/scream/src/mct_coupling/tests/CMakeLists.txt (hunk 5 above)
Comment (bartgol) [reply components/scream/src/mct_coupling/tests/CMakeLists.txt:27]: MPI is required. Not really explicitly, but implicitly. At the beginning of our main CMakeLists.txt, we disable MPI C++ bindings; in order to do this, we try to detect what is the MPI distribution name (openmpi vs mpich, e.g.)., and if we cannot detect the MPI distro name, we error out.

**tl;dr**: you can assume MPI is _always_ available.

File: components/scream/src/mct_coupling/tests/scorpio_tests.cpp [hunk 6]
```diff
@@ -110,5 +119,122 @@ TEST_CASE("scorpio_interface_output", "") {
...
+  register_variable(infilename,"z","z-direction",1,vec_z, PIO_REAL,"z-real");
+  register_variable(infilename,"data_1d","test value for 1d field",1,vec_x, PIO_REAL,"x-real");
+  register_variable(infilename,"data_2d","test value for 2d field",2,vec_xy, PIO_REAL,"xy-real");
+  register_variable(infilename,"data_3d","test value for 3d field",3,vec_xyz, PIO_REAL,"xyz-real");
+  register_variable(infilename,"index_1d","test value for 1d field",1,vec_x, PIO_INT,"x-int");
+  register_variable(infilename,"index_2d","test value for 2d field",2,vec_xy, PIO_INT,"xy-int");
+  register_variable(infilename,"index_3d","test value for 3d field",3,vec_xyz, PIO_INT,"xyz-int");
+
+  // Create data to be written
+  std::array<Real,10> x_data;
```
Comment (PeterCaldwell) [inline components/scream/src/mct_coupling/tests/scorpio_tests.cpp:155]: I'm confused here - this test *reads* rather than writes, right? So the comment above saying data is going to be written is wrong? And isn't the whole point of netcdf format that we can extract the variable info from the .nc file so we don't need to do the register_variable stuff above?

Comment (PeterCaldwell) [reply components/scream/src/mct_coupling/tests/scorpio_tests.cpp:155]: I was originally expecting that this test would write a file, then read it back in. I now think that just reading an existing file is better because it pinpoints that the problem is reading rather than writing. But perhaps we should also change the test from the previous PR to use this read capability to check the output is correct in an entirely self-contained manner?

File: components/scream/src/mct_coupling/scream_scorpio_interface.F90 [hunk 7]
```diff
@@ -342,7 +367,11 @@ subroutine register_variable(pio_atm_filename,shortname,longname,numdims,var_dim
     hist_var%compdof(:my_dof_len) = (/ (ii, ii=istart,istop, 1) /)
 
     ! Register Variable with PIO
-    ierr = PIO_def_var(pio_atm_file%pioFileDesc, trim(shortname), hist_var%dtype, hist_var%dimid(:numdims), hist_var%pioVar)
+    ! check to see if variable already is defined with file (for use with input)
+    ierr = PIO_inq_varid(pio_atm_file%pioFileDesc,trim(shortname),hist_var%piovar)
+    
+    ! if ierr is not pio_noerror then the variable needs to be defined
+    if (ierr.ne.pio_noerr) ierr = PIO_def_var(pio_atm_file%pioFileDesc, trim(shortname), hist_var%dtype, hist_var%dimid(:numdims), hist_var%piovar)
     call errorHandle("PIO ERROR: could not define variable "//trim(shortname),ierr)
```
Comment (PeterCaldwell) [inline components/scream/src/mct_coupling/scream_scorpio_interface.F90:375]: Maybe I don't understand the usage of this function correctly, but I think it is being used for both input and output files, which leads to weirdness. I'd think input files should trigger an error if inq_varid isn't defined (which I don't think is done now) and output files should probably trigger an error if the varid is already defined... Maybe register_variable should only be used for outflds and a "get_variable" function should be added to for use by input reads?

Comment (AaronDonahue) [reply components/scream/src/mct_coupling/scream_scorpio_interface.F90:375]: We could definitely do that to make things clearer. I was hoping to recycle code here since input and output rely on a lot of the same things. But that also means it wouldn’t be hard to create an input specific routine.

File: components/scream/src/mct_coupling/tests/scorpio_tests.cpp (hunk 6 above)
Comment (AaronDonahue) [reply components/scream/src/mct_coupling/tests/scorpio_tests.cpp:155]: The rest test checks the write test output. So if the output is wrong presumably it would be caught here. But we could just as easily do both steps in one test.

Comment (AaronDonahue) [reply components/scream/src/mct_coupling/tests/scorpio_tests.cpp:155]: The rest test checks the write test output. So if the output is wrong presumably it would be caught here. But we could just as easily do both steps in one test.
